src/backends/policy_enforcement/souffle/value_test.cc

```diff
@@ -0,0 +1,172 @@
+
+INSTANTIATE_TEST_SUITE_P(NumberTest, NumberTest,
+                         ValuesIn(kSampleIntegerValues));
+
+class SymbolTest : public TestWithParam<absl::string_view> {};
+
+TEST_P(SymbolTest, SymbolTest) {
+  absl::string_view symbol = GetParam();
+  Symbol symbol_value = Symbol(symbol);
+  std::string symbol_str = symbol_value.ToDatalogString();
+  ASSERT_EQ(symbol_str, "\"" + std::string(symbol) + "\"");
+}
+
+static absl::string_view kSampleSymbols[] = {"", "x", "foo", "hello_world"};
+
+INSTANTIATE_TEST_SUITE_P(SymbolTest, SymbolTest, ValuesIn(kSampleSymbols));
+
+using SimpleRecord = Record<Symbol, Number>;
+
+TEST(SimpleRecordNilTest, SimpleRecordNilTest) {
+  ASSERT_EQ(Record<SimpleRecord>().ToDatalogString(), "nil");
+}
+
+class SimpleRecordTest
+    : public TestWithParam<std::tuple<absl::string_view, int64_t>> {};
+
+TEST_P(SimpleRecordTest, SimpleRecordTest) {
+  const auto [symbol, number] = GetParam();
+  SimpleRecord record_value = SimpleRecord(Symbol(symbol), Number(number));
+  ASSERT_EQ(record_value.ToDatalogString(),
+            absl::StrFormat(R"(["%s", %d])", symbol, number));
+}
+
+INSTANTIATE_TEST_SUITE_P(SimpleRecordTest, SimpleRecordTest,
+                         Combine(ValuesIn(kSampleSymbols),
+                                 ValuesIn(kSampleIntegerValues)));
+
+class NumList : public Record<Number, NumList> {
+ public:
+  using Record::Record;
+};
+
+struct NumListAndExpectedDatalog {
+  const NumList *num_list_ptr;
+  absl::string_view expected_datalog;
+};
+
+class NumListTest : public TestWithParam<NumListAndExpectedDatalog> {};
+
+TEST_P(NumListTest, NumListTest) {
+  const auto [num_list_ptr, expected_datalog] = GetParam();
+  EXPECT_EQ(num_list_ptr->ToDatalogString(), expected_datalog);
+}
+
+static const NumList kEmptyNumList;
+static const NumList kOneElementNumList = NumList(Number(5), NumList());
+static const NumList kTwoElementNumList(Number(-30),
+                                        NumList(Number(28), NumList()));
+
+static NumListAndExpectedDatalog kListAndExpectedDatalog[] = {
+    {.num_list_ptr = &kEmptyNumList, .expected_datalog = "nil"},
+    {.num_list_ptr = &kOneElementNumList, .expected_datalog = "[5, nil]"},
+    {.num_list_ptr = &kTwoElementNumList,
+     .expected_datalog = "[-30, [28, nil]]"}};
+
+INSTANTIATE_TEST_SUITE_P(NumListTest, NumListTest,
+                         ValuesIn(kListAndExpectedDatalog));
+
+static constexpr char kArithAdtName[] = "Arith";
+static constexpr char kNullBranchName[] = "Null";
+static constexpr char kNumberBranchName[] = "Number";
+static constexpr char kAddBranchName[] = "Add";
+
+using ArithAdt = Adt<kArithAdtName>;
+
+class NullBranch : public ArithAdt {
+ public:
+  NullBranch() : Adt(kNullBranchName) {}
+};
+
+class NumberBranch : public ArithAdt {
+ public:
+  NumberBranch(Number number) : Adt(kNumberBranchName) {
+    members_.push_back(std::make_unique<Number>(number));
```

Comment:
Shouldn't you pass this as an argument to the parent's constructor?

Reply:
I've tried that in a few different forms. I always end up running into this error:

```
src/ir/datalog/value_test.cc:133:59:   required from here
/usr/include/c++/11/bits/stl_uninitialized.h:138:72: error: static assertion failed: result type must be constructible from value type of input range
  138 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~
```

GCC (maybe other compilers too?) does not like it when you try to do anything concise or clever with initializing a `std::vector` with an initializer list or another temporary `std::vector` of uncopyable items (like `std::unique_ptr`). I ended up just giving up and doing this ugly thing.

Reply:
SG.